indico/modules/events/registration/client/js/form/fields/DateInput.jsx

```diff
@@ -0,0 +1,113 @@
+// This file is part of Indico.
+// Copyright (C) 2002 - 2021 CERN
+//
+// Indico is free software; you can redistribute it and/or
+// modify it under the terms of the MIT License; see the
+// LICENSE file for more details.
+
+import PropTypes from 'prop-types';
+import TimePicker from 'rc-time-picker';
+import React, {useState} from 'react';
+
+import {SingleDatePicker} from 'indico/react/components';
+import {FinalDropdown} from 'indico/react/forms';
+import {Translate} from 'indico/react/i18n';
+
+export default function DateInput({htmlName, id, disabled, dateFormat, hasTime}) {
+  const [date, setDate] = useState(null);
+  const [time, setTime] = useState(null);
+  const uses24HourFormat = false;
+  const timeDateFormat = dateFormat
+    .replace(/%([HMdmY])/g, (match, c) => ({H: 'HH', M: 'mm', d: 'DD', m: 'MM', Y: 'YYYY'}[c]))
+    .split(' ');
+
+  if (dateFormat.includes('%d')) {
+    return (
+      <>
+        <SingleDatePicker
+          id={`regform-datepicker-${id}`}
+          date={date}
+          onDateChange={setDate}
+          placeholder={timeDateFormat[0]}
+          displayFormat={timeDateFormat[0]}
+          disabled={disabled}
+          isOutsideRange={() => false}
+          enableOutsideDays
+          noBorder
+          small
+        />
+        {hasTime && (
+          <TimePicker
+            id={`regform-timepicker-${id}`}
+            showSecond={false}
+            value={time}
+            focusOnOpen
+            format={timeDateFormat[1]}
+            onChange={setTime}
+            use12Hours={!uses24HourFormat}
+            allowEmpty={false}
+            placeholder="--:--"
+            disabled={disabled}
+            getPopupContainer={node => node}
+            small
+          />
+        )}
+      </>
+    );
+  } else {
+    return (
+      <input type="text" name={htmlName} placeholder={timeDateFormat[0]} disabled={disabled} />
+    );
+  }
+}
+
+DateInput.propTypes = {
+  htmlName: PropTypes.number.isRequired,
+  id: PropTypes.number.isRequired,
+  disabled: PropTypes.bool,
+  dateFormat: PropTypes.string,
+  hasTime: PropTypes.bool,
+};
+
+DateInput.defaultProps = {
+  disabled: false,
+  dateFormat: '%d/%m/%Y',
+  hasTime: false,
+};
+
+export function DateSettings() {
+  const options = [
+    '%d/%m/%Y %H:%M',
+    '%d.%m.%Y %H:%M',
+    '%m/%d/%Y %H:%M',
+    '%m.%d.%Y %H:%M',
+    '%Y/%m/%d %H:%M',
+    '%Y.%m.%d %H:%M',
+    '%d/%m/%Y',
+    '%d.%m.%Y',
+    '%m/%d/%Y',
+    '%m.%d.%Y',
+    '%Y/%m/%d',
+    '%Y.%m.%d',
+    '%m/%Y',
+    '%m.%Y',
+    '%Y',
+  ].map(opt => ({
+    key: opt,
+    value: opt,
+    text: opt.replace(
+      /%([HMdmY])/g,
+      (match, c) => ({H: 'HH', M: 'MM', d: 'DD', m: 'MM', Y: 'YYYY'}[c])
+    ),
+  }));
+  return (
+    <FinalDropdown
+      name="dateFormat"
+      label={Translate.string('Date format')}
+      options={options}
+      defaultValue={DateInput.defaultProps.dateFormat}
```

Comment:
I think using `settingsFormInitialData` in the registry (like done for the choice fields) is better; using a `defaultValue` here makes the field immediately dirty